File: backend/app/services/business_service.py
"""
Business Profile Service

Handles business profile creation and management operations.
"""

import uuid
from datetime import datetime
from sqlalchemy.orm import Session
from sqlalchemy.exc import IntegrityError

from app.models.business_profile import BusinessProfile, CrawlStatus
from app.schemas.business import BusinessProfileCreate


class BusinessService:
    """Service for business profile operations"""

    @staticmethod
    def create_business_profile(
        db: Session,
        user_id: str,
        profile_data: BusinessProfileCreate
    ) -> BusinessProfile:
        """
        Create a new business profile for a user.

        Args:
            db: Database session
            user_id: UUID of the authenticated user
            profile_data: Business profile data from request

        Returns:
            Created BusinessProfile instance

        Raises:
            IntegrityError: If user_id doesn't exist or other DB constraint violated
        """
        # Create business profile instance
        business_profile = BusinessProfile(
            id=str(uuid.uuid4()),
            user_id=user_id,
            name=profile_data.name,
            category=profile_data.category,
            primary_location=profile_data.primary_location,
            website=profile_data.website,
            brand_voice=profile_data.brand_voice,
            main_goal=profile_data.main_goal,
            total_pages_crawled="0",
            created_at=datetime.utcnow(),
            updated_at=datetime.utcnow()
        )
        
        # Set crawl_status using string value to avoid enum serialization issues
        business_profile.crawl_status = "not_started"

        # Add to database
        db.add(business_profile)
        db.commit()
        db.refresh(business_profile)

        return business_profile

    @staticmethod
    def get_user_profiles(db: Session, user_id: str) -> list[BusinessProfile]:
        """
        Get all business profiles for a user.

        Args:
            db: Database session
            user_id: UUID of the user

        Returns:
            List of BusinessProfile instances
        """
        return db.query(BusinessProfile).filter(
            BusinessProfile.user_id == user_id
        ).order_by(BusinessProfile.created_at.desc()).all()

    @staticmethod
    def get_profile_by_id(db: Session, profile_id: str, user_id: str) -> BusinessProfile | None:
        """
        Get a specific business profile by ID, ensuring it belongs to the user.

        Args:
            db: Database session
            profile_id: UUID of the business profile
            user_id: UUID of the authenticated user

        Returns:
            BusinessProfile instance or None if not found
        """
        return db.query(BusinessProfile).filter(
            BusinessProfile.id == profile_id,
            BusinessProfile.user_id == user_id
        ).first()

File: backend/src/services/BusinessService.ts
import { BusinessProfile } from '../models/index.js';
import { NotFoundError, ValidationError } from '../utils/errors.js';
import { Types } from 'mongoose';

export class BusinessService {
  static async createProfile(userId: string, data: {
    name: string;
    category: string;
    primaryLocation: string;
    website?: string;
    brandVoice?: string;
    mainGoal?: string;
  }) {
    if (!data.name || !data.category || !data.primaryLocation) {
      throw new ValidationError('Name, category, and primaryLocation are required');
    }

    const profile = await BusinessProfile.create({
      userId: new Types.ObjectId(userId),
      ...data,
    });

    return profile;
  }

  static async getProfile(profileId: string) {
    const profile = await BusinessProfile.findById(profileId).populate('userId', 'email');
    if (!profile) {
      throw new NotFoundError('Business profile not found');
    }
    return profile;
  }

  static async getProfilesByUser(userId: string) {
    const profiles = await BusinessProfile.find({ userId: new Types.ObjectId(userId) })
      .sort({ createdAt: -1 });
    return profiles;
  }

  static async getCurrentProfile(userId: string) {
    const profile = await BusinessProfile.findOne({ userId: new Types.ObjectId(userId) })
      .sort({ createdAt: -1 });
    return profile;
  }

  static async updateProfile(profileId: string, data: Partial<{
    name: string;
    category: string;
    primaryLocation: string;
    website: string;
    brandVoice: string;
    mainGoal: string;
  }>) {
    const profile = await BusinessProfile.findByIdAndUpdate(profileId, data, { new: true });
    if (!profile) {
      throw new NotFoundError('Business profile not found');
    }
    return profile;
  }

  static async updateCrawlStatus(profileId: string, status: string, error?: string) {
    const updateData: any = { crawlStatus: status };

    if (status === 'IN_PROGRESS') {
      updateData.crawlStartedAt = new Date();
    } else if (status === 'COMPLETED') {
      updateData.crawlCompletedAt = new Date();
    } else if (status === 'FAILED') {
      updateData.crawlError = error;
    }

    const profile = await BusinessProfile.findByIdAndUpdate(profileId, updateData, { new: true });
    if (!profile) {
      throw new NotFoundError('Business profile not found');
    }
    return profile;
  }

  static async deleteProfile(profileId: string) {
    const profile = await BusinessProfile.findByIdAndDelete(profileId);
    if (!profile) {
      throw new NotFoundError('Business profile not found');
    }
    return profile;
  }
}

export default BusinessService;
